File: frontend/src/services/donorService.ts
/**
 * Donor service for API communication
 */

export interface Donor {
  id: number;
  first_name: string;
  last_name: string;
  full_name?: string;
  email?: string;
  phone?: string;
  city?: string;
  state?: string;
  company?: string;
  donor_status: string;
  donor_type: string;
  total_gifts: number;
  total_gift_count: number;
  created_at: string;
}

export interface DonorCreate {
  first_name: string;
  last_name: string;
  email?: string;
  phone?: string;
  mobile_phone?: string;
  work_phone?: string;
  address_line_1?: string;
  address_line_2?: string;
  city?: string;
  state?: string;
  postal_code?: string;
  country?: string;
  company?: string;
  job_title?: string;
  preferred_contact_method?: string;
  do_not_email?: boolean;
  do_not_call?: boolean;
  do_not_mail?: boolean;
  donor_type?: string;
  notes?: string;
  source?: string;
}

export interface DonorUpdate extends Partial<DonorCreate> {
  preferred_name?: string;
  title?: string;
  suffix?: string;
  donor_status?: string;
  wealth_rating?: string;
  capacity_rating?: number;
}

const API_BASE_URL = import.meta.env.VITE_API_URL || 'http://localhost:8000';

class DonorService {
  private baseUrl = `${API_BASE_URL}/api/v1/donors`;

  async createDonor(donor: DonorCreate): Promise<Donor> {
    const response = await fetch(this.baseUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(donor),
    });

    if (!response.ok) {
      throw new Error(`Failed to create donor: ${response.statusText}`);
    }

    return response.json();
  }

  async getDonors(skip = 0, limit = 100): Promise<Donor[]> {
    const url = new URL(this.baseUrl);
    url.searchParams.append('skip', skip.toString());
    url.searchParams.append('limit', limit.toString());

    const response = await fetch(url.toString());

    if (!response.ok) {
      throw new Error(`Failed to fetch donors: ${response.statusText}`);
    }

    return response.json();
  }

  async getDonor(id: number): Promise<Donor> {
    const response = await fetch(`${this.baseUrl}/${id}`);

    if (!response.ok) {
      throw new Error(`Failed to fetch donor: ${response.statusText}`);
    }

    return response.json();
  }

  async updateDonor(id: number, updates: DonorUpdate): Promise<Donor> {
    const response = await fetch(`${this.baseUrl}/${id}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(updates),
    });

    if (!response.ok) {
      throw new Error(`Failed to update donor: ${response.statusText}`);
    }

    return response.json();
  }

  async searchDonors(query: string, limit = 50): Promise<Donor[]> {
    const url = new URL(`${this.baseUrl}/search`);
    url.searchParams.append('q', query);
    url.searchParams.append('limit', limit.toString());

    const response = await fetch(url.toString());

    if (!response.ok) {
      throw new Error(`Failed to search donors: ${response.statusText}`);
    }

    return response.json();
  }

  async findDuplicates(donorId: number): Promise<Donor[]> {
    const response = await fetch(`${this.baseUrl}/${donorId}/duplicates`);

    if (!response.ok) {
      throw new Error(`Failed to find duplicates: ${response.statusText}`);
    }

    return response.json();
  }

  async mergeDonors(primaryId: number, duplicateId: number): Promise<Donor> {
    const response = await fetch(`${this.baseUrl}/merge`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        primary_donor_id: primaryId,
        duplicate_donor_id: duplicateId,
      }),
    });

    if (!response.ok) {
      throw new Error(`Failed to merge donors: ${response.statusText}`);
    }

    return response.json();
  }

  async addTag(donorId: number, tagName: string): Promise<void> {
    const response = await fetch(`${this.baseUrl}/${donorId}/tags`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ tag_name: tagName }),
    });

    if (!response.ok) {
      throw new Error(`Failed to add tag: ${response.statusText}`);
    }
  }

  async deleteDonor(id: number): Promise<void> {
    const response = await fetch(`${this.baseUrl}/${id}`, {
      method: 'DELETE',
    });

    if (!response.ok) {
      throw new Error(`Failed to delete donor: ${response.statusText}`);
    }
  }
}

export const donorService = new DonorService();

// Explicit re-exports to ensure proper module loading
export type { Donor, DonorCreate, DonorUpdate };

File: backend/api/services/donors/donor_service.py
"""Donor service for business logic."""
import re
from typing import List, Optional, Dict, Any
from sqlmodel import Session
from models.donors.donor import Donor
from models.donors.tag import Tag, DonorTag
from repositories.donors.donor_repository import DonorRepository


class DonorService:
    """Service for donor business logic."""
    
    def __init__(self, session: Session):
        self.session = session
        self.repository = DonorRepository(session)
    
    def create_donor(self, donor_data: Dict[str, Any]) -> Donor:
        """Create a new donor with normalized fields."""
        donor = Donor(**donor_data)
        
        # Set full name if not provided
        if not donor.full_name and donor.first_name and donor.last_name:
            donor.full_name = f"{donor.first_name} {donor.last_name}".strip()
        
        # Set normalized keys for duplicate detection
        donor.name_key = self._normalize_name(donor.full_name or "")
        donor.email_key = self._normalize_email(donor.email or "")
        donor.phone_key = self._normalize_phone(donor.phone or "")
        
        return self.repository.create(donor)
    
    def update_donor(self, donor_id: int, donor_data: Dict[str, Any]) -> Optional[Donor]:
        """Update donor with normalized fields."""
        donor = self.repository.get_by_id(donor_id)
        if not donor:
            return None
        
        # Update fields
        for key, value in donor_data.items():
            if hasattr(donor, key):
                setattr(donor, key, value)
        
        # Update full name if first/last name changed
        if "first_name" in donor_data or "last_name" in donor_data:
            if donor.first_name and donor.last_name:
                donor.full_name = f"{donor.first_name} {donor.last_name}".strip()
        
        # Update normalized keys
        donor.name_key = self._normalize_name(donor.full_name or "")
        donor.email_key = self._normalize_email(donor.email or "")
        donor.phone_key = self._normalize_phone(donor.phone or "")
        
        return self.repository.update(donor)
    
    def get_donor(self, donor_id: int) -> Optional[Donor]:
        """Get donor by ID."""
        return self.repository.get_by_id(donor_id)
    
    def list_donors(self, skip: int = 0, limit: int = 100) -> List[Donor]:
        """List donors with pagination."""
        return self.repository.get_all(skip=skip, limit=limit)
    
    def search_donors(self, query: str, limit: int = 50) -> List[Donor]:
        """Search donors by name, email, or phone."""
        return self.repository.search_donors(query, limit)
    
    def find_potential_duplicates(self, donor: Donor) -> List[Donor]:
        """Find potential duplicate donors."""
        return self.repository.find_potential_duplicates(donor)
    
    def merge_donors(self, primary_donor_id: int, duplicate_donor_id: int) -> Optional[Donor]:
        """Merge two donor records."""
        primary = self.repository.get_by_id(primary_donor_id)
        duplicate = self.repository.get_by_id(duplicate_donor_id)
        
        if not primary or not duplicate:
            return None
        
        # Merge logic: keep primary donor's data, but fill in missing fields from duplicate
        merge_fields = [
            'email', 'phone', 'mobile_phone', 'work_phone',
            'address_line_1', 'address_line_2', 'city', 'state', 'postal_code',
            'company', 'job_title', 'notes'
        ]
        
        for field in merge_fields:
            primary_value = getattr(primary, field)
            duplicate_value = getattr(duplicate, field)
            if not primary_value and duplicate_value:
                setattr(primary, field, duplicate_value)
        
        # Update gift totals (would need to move actual gifts in a real implementation)
        primary.total_gifts += duplicate.total_gifts
        primary.total_gift_count += duplicate.total_gift_count
        
        # Update dates
        if duplicate.first_gift_date and (not primary.first_gift_date or duplicate.first_gift_date < primary.first_gift_date):
            primary.first_gift_date = duplicate.first_gift_date
        
        if duplicate.last_gift_date and (not primary.last_gift_date or duplicate.last_gift_date > primary.last_gift_date):
            primary.last_gift_date = duplicate.last_gift_date
        
        # Update largest gift
        if duplicate.largest_gift > primary.largest_gift:
            primary.largest_gift = duplicate.largest_gift
        
        # Recalculate average
        if primary.total_gift_count > 0:
            primary.average_gift = primary.total_gifts / primary.total_gift_count
        
        # Save primary and delete duplicate
        updated_primary = self.repository.update(primary)
        self.repository.delete(duplicate_donor_id)
        
        return updated_primary
    
    def add_tag_to_donor(self, donor_id: int, tag_name: str) -> bool:
        """Add a tag to a donor."""
        donor = self.repository.get_by_id(donor_id)
        if not donor:
            return False
        
        # Find or create tag
        tag_statement = f"SELECT * FROM tags WHERE name = '{tag_name}'"
        existing_tags = self.session.exec(tag_statement).all()
        
        if existing_tags:
            tag = existing_tags[0]
        else:
            tag = Tag(name=tag_name)
            self.session.add(tag)
            self.session.commit()
            self.session.refresh(tag)
        
        # Check if association already exists
        existing_association = self.session.exec(
            f"SELECT * FROM donor_tags WHERE donor_id = {donor_id} AND tag_id = {tag.id}"
        ).first()
        
        if not existing_association:
            donor_tag = DonorTag(donor_id=donor_id, tag_id=tag.id)
            self.session.add(donor_tag)
            self.session.commit()
        
        return True
    
    def _normalize_name(self, name: str) -> str:
        """Normalize name for duplicate detection."""
        if not name:
            return ""
        # Remove extra spaces, convert to lowercase, remove punctuation
        normalized = re.sub(r'[^\w\s]', '', name.lower())
        normalized = ' '.join(normalized.split())
        return normalized
    
    def _normalize_email(self, email: str) -> str:
        """Normalize email for duplicate detection."""
        if not email:
            return ""
        return email.lower().strip()
    
    def _normalize_phone(self, phone: str) -> str:
        """Normalize phone for duplicate detection."""
        if not phone:
            return ""
        # Remove all non-numeric characters
        return re.sub(r'[^\d]', '', phone)